File: rewind/lib/core/CheckpointStore.js
import fs from 'fs/promises';
import path from 'path';
import os from 'os';
import { spawn } from 'child_process';
import { createHash } from 'crypto';
import { atomicWrite, ensureDir, fileExists, safeStatFile } from '../utils/fs-utils.js';
import { ConfigLoader } from './ConfigLoader.js';

/**
 * CheckpointStore - Enhanced file system snapshot manager
 * 
 * Improvements over CheckpointEngine:
 * - Fast scan optimization with persistent signatures
 * - Centralized configuration via ConfigLoader
 * - Better error handling and atomic operations
 * - Diff summaries for UI
 */
export class CheckpointStore {
  constructor({ projectRoot = process.cwd() } = {}) {
    this.projectRoot = path.resolve(projectRoot);
    this.configLoader = new ConfigLoader(projectRoot);
    this.initialized = false;
  }

  /**
   * Initialize storage paths based on configuration
   */
  async initialize() {
    if (this.initialized) return;

    // Use getStorageMode to resolve the correct mode (config > fallback)
    const storageMode = await this.configLoader.getStorageMode();
    
    if (storageMode === 'global') {
      const config = await this.configLoader.getConfig();
      // User-level storage: ~/.rewind/storage/<sanitized_project_path>
      const home = os.homedir();
      const globalRoot = config.storage?.path || path.join(home, '.rewind', 'storage');
      const projectHash = createHash('sha256').update(this.projectRoot).digest('hex').slice(0, 12);
      const projectName = path.basename(this.projectRoot).replace(/[^a-zA-Z0-9_-]/g, '_');
      
      this.baseDir = path.join(globalRoot, `${projectName}_${projectHash}`);
    } else {
      // Project-level storage: <project>/.rewind/code
      this.baseDir = path.join(this.projectRoot, '.rewind', 'code');
    }

    this.snapshotsDir = path.join(this.baseDir, 'snapshots');
    this.changelogFile = path.join(this.baseDir, 'changelog.json');
    this.headSignatureFile = path.join(this.baseDir, 'head_signature');
    
    // Ensure base directory exists (crucial for global mode)
    await ensureDir(this.baseDir);
    this.initialized = true;
  }

  async ensureDirs() {
    if (!this.initialized) await this.initialize();
    await ensureDir(this.snapshotsDir);
  }

  async getConfig() {
    return await this.configLoader.getConfig();
  }

  /**
   * Fast scan: Check if files changed since last checkpoint
   * Uses mtime + size comparison with cached signature
   */
  async hasChanges() {
    try {
      // Get current head signature if it exists
      let headSignature = null;
      if (await fileExists(this.headSignatureFile)) {
        headSignature = await fs.readFile(this.headSignatureFile, 'utf8');
      }

      // If no head signature, assume changes exist
      if (!headSignature) {
        return true;
      }

      // Quick scan: mtime + size only
      const files = await this.scanProject();
      const quickStats = await this.getQuickFileStats(files);
      const quickSignature = this.computeQuickSignature(quickStats);

      return quickSignature !== headSignature;
    } catch (error) {
      // On error, assume changes exist to be safe
      return true;
    }
  }

  /**
   * Get lightweight file stats (mtime + size only)
   */
  async getQuickFileStats(files) {
    const stats = [];
    for (const relPath of files) {
      const stat = await safeStatFile(path.join(this.projectRoot, relPath));
      if (stat) {
        stats.push({
          path: relPath,
          size: stat.size,
          mtimeMs: Math.floor(stat.mtimeMs) // Round to avoid precision issues
        });
      }
    }
    return stats;
  }

  /**
   * Compute quick signature from mtime + size
   */
  computeQuickSignature(quickStats) {
    const hash = createHash('sha256');
    for (const info of quickStats) {
      hash.update(info.path);
      hash.update(String(info.size));
      hash.update(String(info.mtimeMs));
    }
    return hash.digest('hex');
  }

  /**
   * Update the head signature after successful checkpoint
   */
  async updateHeadSignature() {
    try {
      const files = await this.scanProject();
      const quickStats = await this.getQuickFileStats(files);
      const signature = this.computeQuickSignature(quickStats);
      await atomicWrite(this.headSignatureFile, signature, 'utf8');
    } catch (error) {
      // Non-fatal error - checkpoint still succeeded
      console.warn('[CheckpointStore] Failed to update head signature:', error.message);
    }
  }

  /**
   * Scan project files using configuration
   */
  async scanProject() {
    const config = await this.getConfig();
    const gitignorePatterns = await this.configLoader.getGitignorePatterns();
    const shouldIgnore = this.buildIgnoreMatcher(config, gitignorePatterns);

    // Safety: refuse to scan home directory
    const home = os.homedir();
    if (this.projectRoot === home) {
      throw new Error(`Refusing to scan home directory as projectRoot: ${this.projectRoot}`);
    }

    const files = [];

    const walk = async (dir) => {
      const entries = await fs.readdir(dir, { withFileTypes: true });
      for (const entry of entries) {
        const full = path.join(dir, entry.name);
        const rel = path.relative(this.projectRoot, full);
        if (!rel || rel.startsWith('..')) continue;

        if (entry.isDirectory()) {
          if (!shouldIgnore(rel + '/')) {
            await walk(full);
          }
        } else if (entry.isFile()) {
          if (!shouldIgnore(rel)) {
            files.push(rel);
          }
        }
      }
    };

    await walk(this.projectRoot);
    files.sort();
    return files;
  }

  /**
   * Build ignore matcher from configuration
   */
  buildIgnoreMatcher(config, gitignorePatterns) {
    const ignorePatterns = [
      ...(config.ignorePatterns || []),
      ...(config.additionalIgnores || []),
      ...gitignorePatterns
    ];
    const forcePatterns = config.forceInclude || [];

    const ignoreRegexes = ignorePatterns
      .map(p => this.patternToRegex(p))
      .filter(Boolean);
    const forceRegexes = forcePatterns
      .map(p => this.patternToRegex(p))
      .filter(Boolean);

    return relPath => {
      const normalized = relPath.replace(/\\/g, '/');
      
      // Always ignore .rewind directory and its contents
      if (normalized.startsWith('.rewind/') || normalized === '.rewind') {
        return true;
      }
      
      const ignored = ignoreRegexes.some(rx => rx.test(normalized));
      if (!ignored) return false;
      const forceIncluded = forceRegexes.some(rx => rx.test(normalized));
      return !forceIncluded;
    };
  }

  patternToRegex(pattern) {
    const p = pattern.replace(/\\/g, '/').replace(/\/\.$/, '');
    if (!p) return null;
    const escaped = p.replace(/[-/\\^$+?.()|[\]{}]/g, '\\$&').replace(/\*/g, '.*');
    return new RegExp(`^${escaped}$`);
  }

  /**
   * Get detailed file stats for full signature computation
   */
  async getDetailedFileStats(files) {
    const stats = [];
    for (const relPath of files) {
      try {
        const stat = await fs.stat(path.join(this.projectRoot, relPath));
        stats.push({ 
          path: relPath, 
          size: stat.size, 
          mtimeMs: stat.mtimeMs 
        });
      } catch {
        stats.push({ 
          path: relPath, 
          size: 0, 
          mtimeMs: 0 
        });
      }
    }
    return stats;
  }

  computeDetailedSignature(fileStats) {
    const hash = createHash('sha256');
    for (const info of fileStats) {
      hash.update(info.path);
      hash.update(String(info.size));
      hash.update(String(info.mtimeMs));
    }
    return hash.digest('hex');
  }

  /**
   * Create checkpoint with optimization
   */
  async createCheckpoint({ description = '', name, force = false } = {}) {
    await this.ensureDirs();
    
    // Fast check first unless forced
    if (!force && !await this.hasChanges()) {
      return { 
        success: false, 
        noChanges: true, 
        message: 'No changes detected since last checkpoint' 
      };
    }

    const files = await this.scanProject();
    if (!files.length) {
      return { 
        success: false, 
        error: 'No files found to checkpoint', 
        noChanges: true 
      };
    }

    const fileStats = await this.getDetailedFileStats(files);
    const totalBytes = fileStats.reduce((sum, info) => sum + (info.size || 0), 0);
    const signature = this.computeDetailedSignature(fileStats);

    const config = await this.getConfig();
    const effectiveName = this.generateCheckpointName({ name, description, config });
    const snapshotDir = path.join(this.snapshotsDir, effectiveName);
    await ensureDir(snapshotDir);

    const manifest = {
      name: effectiveName,
      timestamp: new Date().toISOString(),
      description: description || 'Rewind checkpoint',
      files,
      fileCount: files.length,
      totalSize: totalBytes,
      signature,
      filesMetadata: fileStats
    };

    const tarPath = path.join(snapshotDir, 'files.tar.gz');
    
    // Create tarball first to catch any errors early
    try {
      await this.createTarball({ files, tarPath });
    } catch (error) {
      // Clean up the snapshot directory if tarball creation fails
      try {
        await fs.rm(snapshotDir, { recursive: true, force: true });
      } catch {
        // Ignore cleanup errors
      }
      throw new Error(`Failed to create checkpoint tarball: ${error.message}`);
    }

    // Only write manifest after successful tarball creation
    await atomicWrite(
      path.join(snapshotDir, 'manifest.json'),
      JSON.stringify(manifest, null, 2),
      'utf8'
    );

    await this.updateHeadSignature();
    await this.cleanupOldCheckpoints();
    await this.logChangelog({
      action: 'CREATE_CHECKPOINT',
      description: `Created checkpoint: ${effectiveName}`,
      details: manifest.description
    });

    return {
      success: true,
      name: effectiveName,
      description: manifest.description,
      fileCount: files.length,
      totalBytes,
      signature
    };
  }

  generateCheckpointName({ name, description, config }) {
    const ts = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
    if (name && name.trim()) {
      const cleanName = name
        .toLowerCase()
        .replace(/[^a-z0-9\s]/g, '')
        .trim()
        .replace(/\s+/g, '_')
        .slice(0, 30) || 'checkpoint';
      return `${cleanName}_${ts}`;
    }
    if (description && description.trim()) {
      const clean = description
        .toLowerCase()
        .replace(/[^a-z0-9\s]/g, '')
        .trim()
        .replace(/\s+/g, '_')
        .slice(0, 30) || 'checkpoint';
      return `${clean}_${ts}`;
    }
    return `checkpoint_${ts}`;
  }

  createTarball({ files, tarPath }) {
    return new Promise((resolve, reject) => {
      const args = ['-czf', tarPath, '-C', this.projectRoot, ...files];
      const proc = spawn('tar', args, { stdio: ['ignore', 'ignore', 'pipe'] });

      let stderr = '';
      proc.stderr.on('data', chunk => {
        stderr += chunk.toString();
      });

      proc.on('close', code => {
        if (code === 0) return resolve();
        const msg = stderr || `tar exited with code ${code}`;
        reject(new Error(`Failed to create tarball: ${msg.trim()}`));
      });
    });
  }

  async listCheckpoints() {
    try {
      await this.ensureDirs();
      const entries = await fs.readdir(this.snapshotsDir, { withFileTypes: true });
      const manifests = [];
      for (const entry of entries) {
        if (!entry.isDirectory()) continue;
        const manifestPath = path.join(this.snapshotsDir, entry.name, 'manifest.json');
        try {
          const raw = await fs.readFile(manifestPath, 'utf8');
          const m = JSON.parse(raw);
          manifests.push(m);
        } catch {
          // Ignore bad manifests
        }
      }
      return manifests.sort((a, b) => new Date(b.timestamp) - new Date(a.timestamp));
    } catch (err) {
      if (err.code === 'ENOENT') return [];
      throw err;
    }
  }

  /**
   * Get diff summary between two checkpoints
   */
  async getDiffSummary(checkpointA, checkpointB) {
    const filesA = new Set(checkpointA.files || []);
    const filesB = new Set(checkpointB.files || []);
    
    const added = [];
    const removed = [];
    const modified = [];

    // Files in B but not A = added
    for (const file of filesB) {
      if (!filesA.has(file)) {
        added.push(file);
      }
    }

    // Files in A but not B = removed
    for (const file of filesA) {
      if (!filesB.has(file)) {
        removed.push(file);
      }
    }

    // Files in both with different metadata = potentially modified
    const metadataA = new Map((checkpointA.filesMetadata || []).map(f => [f.path, f]));
    const metadataB = new Map((checkpointB.filesMetadata || []).map(f => [f.path, f]));

    for (const file of filesA) {
      if (filesB.has(file)) {
        const statA = metadataA.get(file);
        const statB = metadataB.get(file);
        if (statA && statB && 
            (statA.size !== statB.size || statA.mtimeMs !== statB.mtimeMs)) {
          modified.push(file);
        }
      }
    }

    return { added, removed, modified };
  }

  async restoreCheckpoint({ name }) {
    const checkpoints = await this.listCheckpoints();
    if (!checkpoints.length) {
      return { success: false, error: 'No checkpoints available' };
    }

    const target = checkpoints.find(cp => cp.name === name) ||
      checkpoints.find(cp => cp.name.includes(name));

    if (!target) {
      return { success: false, error: `Checkpoint not found: ${name}` };
    }

    // Emergency backup before restore
    const backup = await this.createCheckpoint({ 
      description: 'Auto-backup before restore', 
      name: 'rewind_backup', 
      force: true 
    });
    const backupName = backup.success ? backup.name : null;

    await this.applyRestore(target);
    await this.updateHeadSignature();

    await this.logChangelog({
      action: 'RESTORE_CHECKPOINT',
      description: `Restored checkpoint: ${target.name}`,
      details: backupName ? `Emergency backup: ${backupName}` : 'Emergency backup unavailable'
    });

    return { 
      success: true, 
      restored: target.name, 
      emergencyBackup: backupName 
    };
  }

  async undoLastCheckpoint() {
    const checkpoints = await this.listCheckpoints();
    if (!checkpoints.length) {
      return { success: false, error: 'No checkpoints to undo' };
    }
    return this.restoreCheckpoint({ name: checkpoints[0].name });
  }

  async applyRestore(checkpoint) {
    const currentFiles = await this.scanProject();
    const currentSet = new Set(currentFiles);
    const snapshotSet = new Set(checkpoint.files || []);

    // Delete files that exist now but not in snapshot
    for (const file of currentSet) {
      if (!snapshotSet.has(file)) {
        try {
          await fs.unlink(path.join(this.projectRoot, file));
        } catch {
          // Ignore deletion failures
        }
      }
    }

    // Extract tarball
    const snapshotDir = path.join(this.snapshotsDir, checkpoint.name);
    const tarPath = path.join(snapshotDir, 'files.tar.gz');
    await this.extractTarball({ tarPath });
  }

  extractTarball({ tarPath }) {
    return new Promise((resolve, reject) => {
      const args = ['-xzf', tarPath, '-C', this.projectRoot];
      const proc = spawn('tar', args, { stdio: ['ignore', 'ignore', 'pipe'] });

      let stderr = '';
      proc.stderr.on('data', chunk => {
        stderr += chunk.toString();
      });

      proc.on('close', code => {
        if (code === 0) return resolve();
        const msg = stderr || `tar exited with code ${code}`;
        reject(new Error(`Failed to extract tarball: ${msg.trim()}`));
      });
    });
  }

  async cleanupOldCheckpoints() {
    const config = await this.getConfig();
    const checkpoints = await this.listCheckpoints();
    const now = Date.now();

    const toDelete = [];

    if (config.maxAgeDays && config.maxAgeDays > 0) {
      const cutoff = now - config.maxAgeDays * 24 * 60 * 60 * 1000;
      for (const cp of checkpoints) {
        const ts = new Date(cp.timestamp).getTime();
        if (!Number.isNaN(ts) && ts < cutoff) {
          toDelete.push(cp);
        }
      }
    }

    const remaining = checkpoints.filter(cp => !toDelete.includes(cp));
    if (config.maxCheckpoints && remaining.length > config.maxCheckpoints) {
      const excess = remaining.slice(config.maxCheckpoints);
      toDelete.push(...excess);
    }

    // Delete uniquely
    const unique = Array.from(new Set(toDelete.map(cp => cp.name)));
    for (const name of unique) {
      const dir = path.join(this.snapshotsDir, name);
      try {
        await fs.rm(dir, { recursive: true, force: true });
      } catch {
        // Ignore cleanup failures
      }
    }
  }

  async logChangelog({ action, description, details }) {
    try {
      let entries = [];
      try {
        const raw = await fs.readFile(this.changelogFile, 'utf8');
        entries = JSON.parse(raw);
      } catch (err) {
        if (err.code !== 'ENOENT') throw err;
      }

      entries.unshift({
        timestamp: new Date().toISOString(),
        action,
        description,
        details
      });

      if (entries.length > 50) {
        entries = entries.slice(0, 50);
      }

      await atomicWrite(this.changelogFile, JSON.stringify(entries, null, 2), 'utf8');
    } catch {
      // Changelog failures should not break main flow
    }
  }

  async getLatestCheckpoint() {
    const checkpoints = await this.listCheckpoints();
    return checkpoints.length ? checkpoints[0] : null;
  }
}

File: rewind/src/core/checkpoint_store.py
"""Checkpoint storage for Rewind.

Handles creating, listing, and restoring file snapshots using tarfile.
"""

from __future__ import annotations

import json
import os
import shutil
import tarfile
import tempfile
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Any, Iterator

from ..config.types import IgnoreConfig


@dataclass
class CheckpointMetadata:
    """Metadata for a checkpoint."""
    name: str
    timestamp: str
    description: str
    file_count: int
    total_size: int
    session_id: str | None = None
    has_transcript: bool = False
    transcript: dict[str, Any] | None = None
    
    def to_dict(self) -> dict:
        """Convert to dictionary for JSON serialization."""
        data: dict[str, Any] = {
            "name": self.name,
            "timestamp": self.timestamp,
            "description": self.description,
            "fileCount": self.file_count,
            "totalSize": self.total_size,
            "sessionId": self.session_id,
            "hasTranscript": self.has_transcript,
        }

        if self.transcript is not None:
            data["transcript"] = self.transcript

        return data
    
    @classmethod
    def from_dict(cls, data: dict) -> CheckpointMetadata:
        """Create from dictionary."""
        return cls(
            name=data.get("name", ""),
            timestamp=data.get("timestamp", ""),
            description=data.get("description", ""),
            file_count=data.get("fileCount", 0),
            total_size=data.get("totalSize", 0),
            session_id=data.get("sessionId"),
            has_transcript=data.get("hasTranscript", False),
            transcript=data.get("transcript"),
        )


@dataclass
class CheckpointResult:
    """Result of a checkpoint operation."""
    success: bool
    name: str = ""
    file_count: int = 0
    error: str | None = None


class CheckpointStore:
    """Manages checkpoint storage and retrieval."""
    
    ARCHIVE_NAME = "snapshot.tar.gz"
    METADATA_NAME = "metadata.json"
    
    def __init__(
        self,
        storage_dir: Path,
        project_root: Path,
        ignore_config: IgnoreConfig | None = None,
    ):
        """Initialize checkpoint store.
        
        Args:
            storage_dir: Directory to store checkpoints
            project_root: Project root directory to snapshot
            ignore_config: Patterns for files to ignore
        """
        self.storage_dir = Path(storage_dir)
        self.project_root = Path(project_root)
        self.ignore_config = ignore_config or IgnoreConfig()
        
        # Ensure storage directory exists
        self.storage_dir.mkdir(parents=True, exist_ok=True)
    
    def create(
        self,
        description: str = "",
        session_id: str | None = None,
    ) -> CheckpointResult:
        """Create a new checkpoint.
        
        Args:
            description: Human-readable description
            session_id: Optional session identifier
            
        Returns:
            CheckpointResult with success status and details
        """
        timestamp = datetime.now()
        # Include milliseconds for uniqueness when creating multiple checkpoints quickly
        name = timestamp.strftime("%Y%m%d_%H%M%S") + f"_{timestamp.microsecond // 1000:03d}"
        checkpoint_dir = self.storage_dir / name
        
        try:
            checkpoint_dir.mkdir(parents=True, exist_ok=True)
            
            # Collect files to archive
            files_to_archive = list(self._collect_files())
            if not files_to_archive:
                return CheckpointResult(
                    success=False,
                    error="No files to checkpoint"
                )
            
            # Create tar archive
            archive_path = checkpoint_dir / self.ARCHIVE_NAME
            total_size = 0
            
            with tarfile.open(archive_path, "w:gz") as tar:
                for file_path in files_to_archive:
                    rel_path = file_path.relative_to(self.project_root)
                    tar.add(file_path, arcname=str(rel_path))
                    total_size += file_path.stat().st_size
            
            # Save metadata
            metadata = CheckpointMetadata(
                name=name,
                timestamp=timestamp.isoformat(),
                description=description,
                file_count=len(files_to_archive),
                total_size=total_size,
                session_id=session_id,
            )
            
            metadata_path = checkpoint_dir / self.METADATA_NAME
            with open(metadata_path, "w") as f:
                json.dump(metadata.to_dict(), f, indent=2)
            
            return CheckpointResult(
                success=True,
                name=name,
                file_count=len(files_to_archive),
            )
            
        except Exception as e:
            # Clean up on failure
            if checkpoint_dir.exists():
                shutil.rmtree(checkpoint_dir, ignore_errors=True)
            return CheckpointResult(success=False, error=str(e))
    
    def restore(self, name: str, backup: bool = True) -> CheckpointResult:
        """Restore a checkpoint.
        
        Args:
            name: Checkpoint name to restore
            backup: Whether to create backup before restore
            
        Returns:
            CheckpointResult with success status
        """
        checkpoint_dir = self.storage_dir / name
        archive_path = checkpoint_dir / self.ARCHIVE_NAME
        
        if not archive_path.exists():
            return CheckpointResult(
                success=False,
                error=f"Checkpoint not found: {name}"
            )
        
        try:
            # Create backup if requested
            if backup:
                backup_result = self.create(
                    description=f"Backup before restore to {name}",
                )
                if not backup_result.success:
                    return CheckpointResult(
                        success=False,
                        error=f"Failed to create backup: {backup_result.error}"
                    )
            
            # Extract to temp directory first
            with tempfile.TemporaryDirectory() as tmp_dir:
                tmp_path = Path(tmp_dir)
                
                with tarfile.open(archive_path, "r:gz") as tar:
                    tar.extractall(tmp_path, filter="data")
                
                # Copy files to project root
                file_count = 0
                for root, _, files in os.walk(tmp_path):
                    for file in files:
                        src = Path(root) / file
                        rel_path = src.relative_to(tmp_path)
                        dst = self.project_root / rel_path
                        
                        dst.parent.mkdir(parents=True, exist_ok=True)
                        shutil.copy2(src, dst)
                        file_count += 1
            
            return CheckpointResult(success=True, name=name, file_count=file_count)
            
        except Exception as e:
            return CheckpointResult(success=False, error=str(e))
    
    def list(self) -> list[CheckpointMetadata]:
        """List all checkpoints.
        
        Returns:
            List of checkpoint metadata, sorted by timestamp (newest first)
        """
        checkpoints = []
        
        if not self.storage_dir.exists():
            return checkpoints
        
        for entry in self.storage_dir.iterdir():
            if not entry.is_dir():
                continue
            
            metadata_path = entry / self.METADATA_NAME
            if metadata_path.exists():
                try:
                    with open(metadata_path) as f:
                        data = json.load(f)
                    checkpoints.append(CheckpointMetadata.from_dict(data))
                except (OSError, json.JSONDecodeError):
                    # Create minimal metadata from directory name
                    checkpoints.append(CheckpointMetadata(
                        name=entry.name,
                        timestamp=entry.name,
                        description="",
                        file_count=0,
                        total_size=0,
                    ))
        
        # Sort by name (which is timestamp-based) descending
        checkpoints.sort(key=lambda x: x.name, reverse=True)
        return checkpoints
    
    def get(self, name: str) -> CheckpointMetadata | None:
        """Get metadata for a specific checkpoint.
        
        Args:
            name: Checkpoint name
            
        Returns:
            CheckpointMetadata or None if not found
        """
        metadata_path = self.storage_dir / name / self.METADATA_NAME
        if not metadata_path.exists():
            return None
        
        try:
            with open(metadata_path) as f:
                return CheckpointMetadata.from_dict(json.load(f))
        except (OSError, json.JSONDecodeError):
            return None
    
    def delete(self, name: str) -> bool:
        """Delete a checkpoint.
        
        Args:
            name: Checkpoint name to delete
            
        Returns:
            True if deleted successfully
        """
        checkpoint_dir = self.storage_dir / name
        if not checkpoint_dir.exists():
            return False
        
        try:
            shutil.rmtree(checkpoint_dir)
            return True
        except OSError:
            return False
    
    def prune(self, keep: int = 10) -> int:
        """Prune old checkpoints, keeping the most recent.
        
        Args:
            keep: Number of checkpoints to keep
            
        Returns:
            Number of checkpoints deleted
        """
        checkpoints = self.list()
        if len(checkpoints) <= keep:
            return 0
        
        to_delete = checkpoints[keep:]
        deleted = 0
        for cp in to_delete:
            if self.delete(cp.name):
                deleted += 1
        
        return deleted
    
    def update_metadata(self, name: str, **updates) -> bool:
        """Update checkpoint metadata.
        
        Args:
            name: Checkpoint name
            **updates: Fields to update
            
        Returns:
            True if updated successfully
        """
        metadata = self.get(name)
        if not metadata:
            return False
        
        # Update fields
        for key, value in updates.items():
            if hasattr(metadata, key):
                setattr(metadata, key, value)
        
        # Save
        metadata_path = self.storage_dir / name / self.METADATA_NAME
        try:
            with open(metadata_path, "w") as f:
                json.dump(metadata.to_dict(), f, indent=2)
            return True
        except OSError:
            return False
    
    def _collect_files(self) -> Iterator[Path]:
        """Collect files to include in checkpoint.
        
        Yields:
            Paths to files that should be checkpointed
        """
        for root, dirs, files in os.walk(self.project_root):
            root_path = Path(root)
            rel_root = root_path.relative_to(self.project_root)
            
            # Filter directories in-place to skip ignored ones
            dirs[:] = [
                d for d in dirs
                if not self.ignore_config.should_ignore(str(rel_root / d))
            ]
            
            for file in files:
                rel_path = rel_root / file
                if not self.ignore_config.should_ignore(str(rel_path)):
                    yield root_path / file
